enhancements/microshift/microshift-supportability-tools.md

```diff
@@ -0,0 +1,297 @@
+1. Expects certain OpenShift components to be present, like machine-config-daemon.
+1. Runs commands within a pod, which means pulling another container over network (or pre-caching it).
+1. Pulls info from many components not present in MicroShift.
+
+Some of these reasons are a no-go for MicroShift because of constrained
+environments. Some others mean there should be significant changes to
+`must-gather` to have MicroShift awareness.
+
+There is another tool for gathering info from a running RHEL system, and that
+is [`sos`](https://sos.readthedocs.org/). This tool is present in RHEL repos
+and is capable of generating reports for many components and applications in
+the system by using plugins. By executing `sos report` it will generate a
+detailed report with all the available information it can extract.
+
+Plugins have an enablement check or may be invoked manually when executing
+the command. A [plugin](https://github.com/sosreport/sos/wiki/How-to-Write-a-Plugin)
+is a python script that we include as part of `sos`.
+
+There is a [plugin for OpenShift](https://github.com/sosreport/sos/blob/main/sos/report/plugins/openshift.py)
+already that almost fits the needs of this EP. MicroShift does not share the
+same resources, APIs and namespaces/names that the plugin assumes, as it was
+made for OpenShift clusters. However, much of the needs covered in the plugin
+are the same, so the contents of the plugin may be used too.
+
+A simple `sos report` will yield much of the information needed to debug a
+MicroShift deployment (the list of active plugins can be obtained by running
+`sos report -l`). However, there are some specifics that we need to add to a
+new plugin:
+1. Control plane component logs: apiserver, etcd, kubelet, etc. (these are
+retrieved by syslog, meaning they are mixed with other system logs).
+1. Yaml output for relevant manifests from API server.
+1. Logs for every container. These are already retrieved by CRIO plugin. May
+need easier indexing.
+1. MicroShift's configuration and version. This [enhancement](https://github.com/openshift/enhancements/pull/1304)
+might add additional information to gather: plugins, manifests, etc.
+
+Retrieving all this information involves using a kubernetes client. While
+MicroShift itself uses the go client extensively, the idea of introducing
+another command to produce this output and avoid installing `oc`/`kubectl` is
+not the best, since we would be adding runtime dependencies between a system
+component and MicroShift.
+
+### Workflow Description
+
+> _Disclaimer:_ Workflows described here illustrate manual processes.
+
+Assuming from now on MicroShift has already been installed in a R4E/RHEL based
+system and there has been some kind of issue.
+
+1. MicroShift administrator notices an issue/wants to get a system report.
+1. The administrator will log in to the failing host.
+1. The administrator will perform the debug report creation procedure:
+```bash
+sudo sos report
+```
```

Comment:
In my tests using the demos, root has no exposure to a kube config file default. I've been using "sudo sos report -v -o  openshift -k openshift.with-api -k openshift.kubeconfig=/home/microshift/.kube/config"  variations with "-k openshift.kubeconfig" to point at a location. So maybe this would have to be called out in docs or the microshift plugin could use a different default kube config location.

Reply:
It could also be /var/lib/microshift/resources/kubeadmin/kubeconfig

Reply:
Good point. The microshift plugin can have built into it the knowledge that we put the kubeconfig in /var/lib/microshift/resources/kubeadmin/kubeconfig

Reply:
The kubeconfig in `/var/lib/microshift/resources/kubeadmin/kubeconfig` is always present and works under the assumption of using localhost. I added this to my working prototype of the plugin. Will post a draft ASAP.

Reply:
You can use token instead. https://github.com/sosreport/sos/blob/main/sos/report/plugins/openshift.py#L132